server/webhook.go

```diff
@@ -243,6 +235,74 @@ func (s *Server) registerWebhookRoutes(g *echo.Group) {
 	})
 }
 
+// We are observing the push webhook event so that we will receive the event either when:
+// 1. A commit is directly pushed to a branch.
+// 2. One or more commits are merged to a branch.
+//
+// There is a complication to deal with the 2nd type. A typical workflow is a developer first
+// commits the migration file to the feature branch, and at a later point, she creates a merge
+// request to merge the commit to the main branch. Even the developer only creates a single commit
+// on the feature branch, that merge request may contain multiple commits (unless both squash and fast-forward merge are used):
+// 1. The original commit on the feature branch.
+// 2. The merge request commit.
+//
+// And both commits would include that added migration file. Since we create an issue per migration file,
+// we need to filter the commit list to prevent creating a duplicated issue. GitLab has a limitation to distinguish
+// whether the commit is a merge commit (https://gitlab.com/gitlab-org/gitlab/-/issues/30914), so we need to dedup
+// ourselves. Below is the filtering algorithm:
+// 1. If we observe the same migration file multiple times, then we should use the latest migration file. This does not matter
+//    for change-based migration since a developer would always create different migration file with incremental names, while it
+//    will be important for the state-based migration, since the file name is always the same and we need to use the latest snapshot.
+// 2. Maintain the relative commit order between different migration files. If migration file A happens before migration file B,
+//    then we should create an issue for migration file A first.
+type distinctFileItem struct {
+	createdTime time.Time
+	commit      gitlab.WebhookCommit
+	fileName    string
+}
+
+func dedupMigrationFilesFromCommitList(commitList []gitlab.WebhookCommit) []distinctFileItem {
+	distinctFileList := []distinctFileItem{}
+	for _, commit := range commitList {
+		log.Debug("Pre-processing commit to dedup migration files...",
+			zap.String("id", common.EscapeForLogging(commit.ID)),
+			zap.String("title", common.EscapeForLogging(commit.Title)),
+		)
+
+		createdTime, err := time.Parse(time.RFC3339, commit.Timestamp)
+		if err != nil {
+			log.Warn("Ignored commit, failed to parse commit timestamp.", zap.String("commit", common.EscapeForLogging(commit.ID)), zap.String("timestamp", common.EscapeForLogging(commit.Timestamp)), zap.Error(err))
+		}
+
+		for _, added := range commit.AddedList {
+			new := true
+			for i, file := range distinctFileList {
```

Comment:
How about using `distinctFileList  := map[string]distinctFileItem` (filename->distinctFileItem) to make this block of code more readable?
```
file, exist := distinctFileList[added]
if exist { ... } else { ... }
```

Reply:
we'd like to maintain order.

If the feature branch contains 2 commits, 1 commit changes file1.sql, 1 commit changes file2.sql, we also want to create an issue for file1.sql before creating issue for file2.sql.

Reply:
SG.